env/env.go

```diff
@@ -0,0 +1,141 @@
+package env
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+)
+
+// Lookup must return the value for the given name and whether or not it was
+// found.
+type Lookup func(name string) (string, bool)
+
+// VarSet provides methods to access environment variables from a list of Lookup
+// sources.
+type VarSet struct {
+	// sources are searched sequentially for a given environment variable.
+	sources []Lookup
+}
+
+// NewVarSet makes a new VarSet from opts.
+func NewVarSet(opts ...VarSetOpt) *VarSet {
+	v := &VarSet{sources: make([]Lookup, 0)}
+	for _, opt := range opts {
+		opt(v)
+	}
+	return v
+}
+
+// AddSource adds s as another lookup source.
+func (v *VarSet) AddSource(s Lookup) *VarSet {
+	v.sources = append(v.sources, s)
+	return v
+}
+
+// String gets the string environment variable with the given name, if set.
+// If not found, returns false.
+func (v *VarSet) String(name, defaultVal string) string {
+	rv, ok := v.load(name)
+	if !ok {
+		return defaultVal
+	}
+	return rv
+}
+
+// MustString gets the string environment variable with the given name, if set.
+// It will panic with an error if it is not set.
+// Callers can use IsVarNotFound to check the error.
+func (v *VarSet) MustString(name string) string {
+	rv, ok := v.load(name)
+	if !ok {
+		panic(&varNotFoundError{name: name})
+	}
+	return rv
+}
+
+// Bool gets the boolean environment variable with the given name, if set.
+// If not found, returns false.
+// If found and cannot parse, returns an error.
+func (v *VarSet) Bool(name string) (bool, error) {
+	val, ok := v.load(name)
+	if !ok {
+		return false, nil
+	}
+	rv, err := strconv.ParseBool(val)
+	if err != nil {
+		return false, err
+	}
+	return rv, nil
+}
+
+// MustBool gets the boolean environment variable with the given name, if set.
+// It will panic with an error if it is not set or is not a valid boolean.
```

Comment:
So that wasn't quite the intention of the original method, ie if not set, I don't want a panic, I just want `false` - usually a bool is being set in config to override an otherwise sensible settling like "I want secure cookies" - normally it won't be set, but if someone has gone to the effort of setting it, they'd like to know if they've typoed it.

Not sure what that method should be called though....

Reply:
If I've understood correctly, I've left the same behaviour, except with a slightly extra step where you now just need to check the error coming from `Bool()`. E.g. usage here: https://github.com/18F/cg-dashboard/pull/1258/files#diff-661525df772e761479305d75690a20adR125

In that case, if the env var is not set, you just get false. If it is set, we check that it parses and if not you'll get an error.

The confusion arises because we have a few notions to mix together and we have to define what they mean. E.g. does `Must` mean that it must exist or must parse? It could mean either (e.g. `html/template.MustParse` means parse). However `env.MustString` means it must exist. So to be consistent in the `env` package, I think that `env.MustBool` should mean that it must exist - and if it must exist by definition it must also parse so that case is covered too. The only other case left is `MustParseBool` which would mean "if the env var does not exist, return false. if the env var does exist, try and parse it. if it does not parse, then panic. if it does parse, return it".

Summary: we already have the behaviour of `env.MustParseBool` in `env.Bool` except the caller just needs to check the error. I can add this though if you think it sounds OK?